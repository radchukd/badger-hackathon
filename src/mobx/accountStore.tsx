import { action, extendObservable } from 'mobx';
import fetch from 'node-fetch';
import testAccount from '../components/__tests__/testAccount.json';

import { Account } from '../model/account.interface';
import { SettBalance } from '../model/sett-balance.interface';
import { SettVault } from '../model/sett-vault.interface';

export type AccountStoreType = {
  preload: boolean;
};

export class AccountStore {
  private baseUrl = 'https://api.badger.finance/v2';
  public account?: Account | null;
  public prices?: Record<string, number>;
  public vaults?: Array<SettVault>;
  public lastUpdate: Date | null = null;

  constructor({ preload }: AccountStoreType) {
    extendObservable(this, {
      account: this.account,
      prices: this.prices,
      vaults: this.vaults,
    });

    if (preload) {
      this.loadAccount();
      this.loadPrices();
      this.loadVaults();
      this.lastUpdate = new Date();
    }
  }

  loadAccount = action(async (): Promise<void> => {
    //const res = await fetch(`${this.baseUrl}/accounts/0x4e65175f05b4140a0747c29cce997cd4bb7190d4`);

    //if (res.ok) this.account = await res.json();
    //else this.account = null;
    this.account = testAccount;
  });

  loadPrices = action(async (): Promise<void> => {
    const res = await fetch(`${this.baseUrl}/prices`);

    if (res.ok) this.prices = await res.json();
    else this.prices = {};
  });

  loadVaults = action(async (): Promise<void> => {
    const res = await fetch(`${this.baseUrl}/setts`);

    if (res.ok) this.vaults = await res.json();
    else this.vaults = [];
  });

  get isLoading(): boolean {
    return (
      typeof this.account === 'undefined' || typeof this.prices === 'undefined' || typeof this.vaults === 'undefined'
    );
  }

  get roiPercentage(): number {
    if (!this.account?.value || !this.account?.earnedValue) return 0;

    return this.account?.earnedValue / this.account?.value;
  }

  get totalClaimable(): number {
    return this.account?.claimableBalances?.reduce((prev, cur) => prev + (cur?.balance ?? 0), 0) ?? 0;
  }

  get earnedBadger(): number {
    return this.account?.balances?.find((balance) => balance.asset === 'BADGER')?.earnedValue ?? 0;
  }

  assetValue = (balance: SettBalance): number => {
    return balance.balance ? balance.value / balance.balance : 0;
  };

  assetValueBTC = (balance: SettBalance): number => {
    if (!this.prices) return 0;

    const btcPrice = this.prices['0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599'];

    if (!btcPrice) return 0;

    return this.assetValue(balance) / btcPrice;
  };

  allocationDistribution = (type: 'balance' | 'strategy'): Array<{ name: string; value: number }> => {
    const totalAllocation = this.totalAllocation(true, type);

    if (!totalAllocation) return [];

    const top4 =
      this.account?.balances
        ?.slice()
        ?.sort((b1, b2) =>
          type === 'balance'
            ? Math.abs(b2.value) - Math.abs(b1.value)
            : Math.abs(b2.earnedValue) - Math.abs(b1.earnedValue),
        )
        ?.slice(0, 4)
        ?.map(({ asset, value, earnedValue }) => ({
          name: asset,
          value: Math.abs((type === 'balance' ? value : earnedValue) / totalAllocation),
        })) ?? [];
    const top4total = top4?.reduce((prev, cur) => prev + cur.value ?? 0, 0);
    const other = { name: 'Other', value: 1 - top4total };

    return [...top4, other];
  };

  allocationPerAsset = (value: number, type: 'balance' | 'strategy'): number => {
    const totalAllocation = this.totalAllocation(true, type);

    if (!totalAllocation) return 0;

    return Math.abs(value / totalAllocation);
  };

  totalAllocation = (abs: boolean, type: 'balance' | 'strategy'): number => {
    return (
      this.account?.balances?.reduce((prev, cur) => {
        const curValue = (type === 'balance' ? cur.value : cur.earnedValue) ?? 0;
        return prev + (abs ? Math.abs(curValue) : curValue);
      }, 0) ?? 0
    );
  };

  strategyROI = (asset: string): number => {
    if (!this.vaults) return 0;

    const vault = this.vaults.find((vault) => vault.asset === asset);

    if (!vault) return 0;

    return vault.apr;
  };
}

/*
declare let window: Window & { ethereum };

if (!window.ethereum) {
    console.log('Non-Ethereum browser detected.');
  }

  const addresses: Array<string> = await window.ethereum.request({
    method: 'eth_requestAccounts',
  });

  if (addresses.length < 0) {
    console.log('No account found.');
  }

  const web3 = new Web3(Web3.givenProvider);

  addresses.forEach(async (address) => {
    const res = await fetch(`${this.baseUrl}/accounts/${address}`);

    if (res.ok) {
      this.account = await res.json();
      throw new Error('Account is found.');
    }
  });
*/
