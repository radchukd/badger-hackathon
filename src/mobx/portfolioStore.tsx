import { action, makeAutoObservable } from 'mobx';

export type Vault = { id: number; value: string; tags: Array<string> };

export class PortfolioStore {
  public selectedVault?: Vault;
  public vaults: Array<Array<Vault>> = [
    [
      { id: 0, value: 'wBTC/Digg', tags: ['UNI'] },
      { id: 1, value: 'Badger/wBTC', tags: ['UNI'] },
    ],
    [
      { id: 2, value: 'Wrapped BTC/Digg', tags: ['SUSHI'] },
      { id: 3, value: 'Wrapped BTC/Badger', tags: ['SUSHI'] },
      { id: 4, value: 'Wrapped BTC/Wrapped Ether', tags: ['SUSHI'] },
    ],
    [
      { id: 5, value: 'crvRenWBTC', tags: ['CURVE'] },
      { id: 6, value: 'renBTC/wBTC/sBTC', tags: ['CURVE'] },
      { id: 7, value: 'tBTC/sBTCCrv LP', tags: ['CURVE'] },
      { id: 8, value: 'crvRenWBTC', tags: ['CURVE', 'HARVEST'] },
    ],
  ];
  public selectedVaults: Array<number> = this.vaults.flat().map((vault) => vault.id);

  constructor() {
    makeAutoObservable(this);
  }

  setSelectedVault = action((vault: string | undefined): void => {
    if (vault === undefined) {
      this.selectedVault = undefined;
    } else {
      const normalizedName = vault.replace('Sushiswap ', '').replace('Convex ', '');
      this.selectedVault = this.vaults.flat().find((v) => v.value === normalizedName);
    }
  });

  get areAllSelected(): boolean {
    return this.vaults.flat().length === this.selectedVaults.length;
  }

  isVaultSelected = (vault: Vault): boolean => {
    return this.selectedVaults.includes(vault.id);
  };

  onAllClick = action((): void => {
    if (this.areAllSelected) this.selectedVaults = [];
    else this.selectedVaults = this.vaults.flat().map((vault) => vault.id);
  });

  onVaultClick = action((vault: Vault): void => {
    this.selectedVaults = this.isVaultSelected(vault)
      ? this.selectedVaults.filter((selectedVaultId) => selectedVaultId !== vault.id)
      : [...this.selectedVaults, vault.id];
  });
}
