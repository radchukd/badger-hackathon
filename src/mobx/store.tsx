import { action, extendObservable } from 'mobx';
import { RouterStore } from 'mobx-router';
import fetch from 'node-fetch';
import { Account } from '../model/account.interface';
import { SettBalance } from '../model/sett-balance.interface';
import { SettVault } from '../model/sett-vault.interface';

//declare let window: Window & { ethereum };

export class RootStore {
  private baseUrl = 'https://api.badger.finance/v2';
  public router: RouterStore<RootStore>;
  public account?: Account;
  public prices?: Record<string, number>;
  public vaults?: SettVault[];

  constructor() {
    this.router = new RouterStore<RootStore>(this);

    extendObservable(this, {
      account: this.account,
      prices: this.prices,
      vaults: this.vaults,
    });

    this.loadAccount();
    this.loadPrices();
    this.loadVaults();
  }

  loadAccount = action(async (): Promise<void> => {
    /* if (!window.ethereum) {
      console.log('Non-Ethereum browser detected.');
    }

    const addresses: Array<string> = await window.ethereum.request({
      method: 'eth_requestAccounts',
    });

    if (addresses.length < 0) {
      console.log('No account found.');
    }

    const web3 = new Web3(Web3.givenProvider);

    addresses.forEach(async (address) => {
      const res = await fetch(`${this.baseUrl}/accounts/${address}`);

      if (res.ok) {
        this.account = await res.json();
        throw new Error('Account is found.');
      }
    }); */

    const res = await fetch(`${this.baseUrl}/accounts/0x4e65175f05b4140a0747c29cce997cd4bb7190d4`);
    if (res.ok) {
      this.account = await res.json();
    }
  });

  loadPrices = action(async (): Promise<void> => {
    const res = await fetch(`${this.baseUrl}/prices`);
    if (res.ok) {
      this.prices = (await res.json()) || [];
    }
  });

  loadVaults = action(async (): Promise<void> => {
    const res = await fetch(`${this.baseUrl}/setts`);
    if (res.ok) {
      this.vaults = (await res.json()) || [];
    }
  });

  get isLoading(): boolean {
    return (
      typeof this.account === 'undefined' || typeof this.prices === 'undefined' || typeof this.vaults === 'undefined'
    );
  }

  get roiPercentage(): number {
    if (!this.account?.value || !this.account?.earnedValue) return 0;
    return this.account?.earnedValue / this.account?.value;
  }

  get totalClaimable(): number {
    return this.account?.claimableBalances?.reduce((prev, cur) => prev + (cur?.balance ?? 0), 0) ?? 0;
  }

  get earnedBadger(): number {
    if (!this.prices) return 0;

    const badgerPrice = this.prices['0x3472A5A71965499acd81997a54BBA8D852C6E53d'];

    if (!badgerPrice || !this.account?.earnedValue) return 0;

    return this.account?.earnedValue / badgerPrice;
  }

  assetValue = (balance: SettBalance): number => {
    return balance.balance ? balance.value / balance.balance : 0;
  };

  assetValueBTC = (balance: SettBalance): number => {
    if (!this.prices) return 0;

    const btcPrice = this.prices['0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599'];

    if (!btcPrice) return 0;

    return this.assetValue(balance) / btcPrice;
  };

  allocationDistribution = (type: 'balance' | 'strategy'): Array<{ name: string; value: number }> => {
    const totalAllocation = this.totalAllocation(true, type);

    if (!totalAllocation) return [];

    const top4 =
      this.account?.balances
        ?.slice()
        ?.sort((b1, b2) =>
          type === 'balance'
            ? Math.abs(b2.value) - Math.abs(b1.value)
            : Math.abs(b2.earnedValue) - Math.abs(b1.earnedValue),
        )
        ?.slice(0, 4)
        ?.map(({ asset, value, earnedValue }) => ({
          name: asset,
          value: Math.abs((type === 'balance' ? value : earnedValue) / totalAllocation),
        })) ?? [];
    const top4total = top4?.reduce((prev, cur) => prev + cur.value ?? 0, 0);
    const other = { name: 'Other', value: 1 - top4total };
    return [...top4, other];
  };

  allocationPerAsset = (value: number, type: 'balance' | 'strategy'): number => {
    const totalAllocation = this.totalAllocation(true, type);

    if (!totalAllocation) return 0;

    return Math.abs(value / totalAllocation);
  };

  totalAllocation = (abs: boolean, type: 'balance' | 'strategy'): number => {
    return (
      this.account?.balances?.reduce((prev, cur) => {
        const curValue = (type === 'balance' ? cur.value : cur.earnedValue) ?? 0;
        return prev + (abs ? Math.abs(curValue) : curValue);
      }, 0) ?? 0
    );
  };

  strategyROI = (asset: string): number => {
    if (!this.vaults) return 0;

    const vault = this.vaults.find((vault) => vault.asset === asset);

    if (!vault) return 0;

    return vault.apr;
  };
}

const store = new RootStore();

export default store;
