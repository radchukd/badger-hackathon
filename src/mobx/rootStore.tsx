import { RouterStore } from 'mobx-router';
import React from 'react';

import { AccountStore } from './accountStore';
import { PortfolioStore } from './portfolioStore';

export class RootStore {
  public router: RouterStore<RootStore>;

  constructor() {
    this.router = new RouterStore<RootStore>(this);
  }
}

export const rootStore = new RootStore();

export const accountStore = new AccountStore({ preload: true });

export const portfolioStore = new PortfolioStore();

export interface IStoreContext {
  rootStore: RootStore;
  accountStore: AccountStore;
  portfolioStore: PortfolioStore;
}

export const StoreContext = React.createContext({} as IStoreContext);

export const StoreProvider = StoreContext.Provider;
