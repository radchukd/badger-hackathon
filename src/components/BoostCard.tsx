import clsx from 'clsx';
import { observer } from 'mobx-react-lite';
import React from 'react';

import { Card, CardContent, makeStyles, Typography } from '@material-ui/core';

import { StoreContext } from '../mobx/rootStore';
import useCardStyles from '../styles/cardStyles';
import { formatNumber } from '../utils/numberUtils';

const useStyles = makeStyles((theme) => ({
  rank: {
    marginTop: theme.spacing(2),
    fontSize: '16px',
    lineHeight: '28px',
  },
}));

const BoostCard = observer(() => {
  const classes = { ...useStyles(), ...useCardStyles() };
  const store = React.useContext(StoreContext);
  const {
    accountStore: { account },
  } = store;

  return (
    <Card className={classes.cardRoot}>
      <CardContent className={clsx(classes.content, classes.centeredContent)}>
        <Typography className={classes.cardHeading1}>{formatNumber(account?.boost, 'decimal')}</Typography>
        <Typography className={classes.cardSubheading1}>My Boost</Typography>
        <Typography className={classes.rank}>Rank No. {account?.boostRank}</Typography>
      </CardContent>
    </Card>
  );
});

export default BoostCard;
