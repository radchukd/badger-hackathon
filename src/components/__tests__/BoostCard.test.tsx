import '@testing-library/jest-dom';

import React from 'react';

import { render } from '@testing-library/react';

import { RootStore, StoreProvider } from '../../mobx/store';
import BoostCard from '../BoostCard';
import testAccount from './testAccount.json';

describe('BoostCard', () => {
  const store = new RootStore({ preload: false });
  store.account = testAccount;

  it('renders correctly', () => {
    const tree = render(
      <StoreProvider value={store}>
        <BoostCard />
      </StoreProvider>,
    );

    expect(tree).toMatchSnapshot();
  });
});
