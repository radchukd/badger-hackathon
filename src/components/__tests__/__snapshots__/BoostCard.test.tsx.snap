// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoostCard renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardRoot-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root makeStyles-content-3 makeStyles-centeredContent-4"
        >
          <p
            class="MuiTypography-root makeStyles-cardHeading1-5 MuiTypography-body1"
          >
            2.65
          </p>
          <p
            class="MuiTypography-root makeStyles-cardSubheading1-6 MuiTypography-body1"
          >
            My Boost
          </p>
          <p
            class="MuiTypography-root makeStyles-rank-1 MuiTypography-body1"
          >
            Rank No. 
            742
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-cardRoot-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root makeStyles-content-3 makeStyles-centeredContent-4"
      >
        <p
          class="MuiTypography-root makeStyles-cardHeading1-5 MuiTypography-body1"
        >
          2.65
        </p>
        <p
          class="MuiTypography-root makeStyles-cardSubheading1-6 MuiTypography-body1"
        >
          My Boost
        </p>
        <p
          class="MuiTypography-root makeStyles-rank-1 MuiTypography-body1"
        >
          Rank No. 
          742
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
